### PR TITLE
Final 2.0 schema

### PR DESCRIPTION
**Pull request by rjeberhard**
During our review process, we found several issues where either: a) agreed to and spec'd 2.0 functionality was missing or incomplete, or b) 1.x functionality and schema items that were dead had not yet been removed.

Major points:
- Correction of the updated service label, annotation, and nodePort model (serverService, clusterService, and channels)
- Removal of 1.x exportedNetworkAccessPoints
- Correction of serverStartPolicy with the intended enum choices at the right scopes.
- Lots of cleaned-up descriptions and required vs. non-required designations.

### CODE REVIEW

**Comment by rjeberhard** [comment]
@rosemarymarano What's the best way to review the "@Description" values?

File: docs/domains/Domain.json [hunk 1]
```diff
@@ -54,16 +44,30 @@
...
       "properties": {
         "channels": {
-          "$ref": "#/definitions/Map"
-        },
-        "annotations": {
-          "$ref": "#/definitions/Map"
-        },
-        "labels": {
-          "$ref": "#/definitions/Map"
+          "description": "Specifies which of the admin server\u0027s WebLogic channels should be exposed outside the Kubernetes cluster via a node port service, along with the node port for each channel. If not specified, the admin server\u0027s node port service will not be created.",
```
Comment (tbarnes-us) [inline docs/domains/Domain.json]: What is \u0027s?

**Comment by tbarnes-us** [comment]
How do we view the generated output for the Domain.json text?

**Comment by rjeberhard** [comment]
@rosemarymarano The Domain.json file is generated based on the Java; however, there are some repetitions if the same type (i.e. Java class) fits into the overall schema multiple places, such as the KubernetesResource or BaseConfiguration types showing up at the DomainSpec, Cluster, and AdminServer/ManagedServer levels.

There are only a handful of Java classes though, so we could alternatively do an interactive review.

@tbarnes-us The version of Domain.json _on the master branch_ shows up here: https://oracle.github.io/weblogic-kubernetes-operator/domains/Domain.json.  You can also just open the file in a browser.  We also generate a swagger version (that includes the operations you can do on the Domain resource), which is docs/domain/index.html.  Again, the version on oracle.github.io is from the master branch.

File: docs/domains/index.html [hunk 2]
```diff
@@ -1143,6 +1152,10 @@
           "description": "The name of the configmap for optional WebLogic configuration overrides.",
```
Comment (rosemarymarano) [inline docs/domains/index.html]: configmap -> config map

File: model/src/test/resources/oracle/kubernetes/weblogic/domain/v2/domain-sample.yaml [hunk 3]
```diff
@@ -43,14 +43,6 @@ spec:
...
 
-  # Configured storage using nfs style
-  storage:
-    generated:
-      storageSize: 8Gi
-      nfs:
-        server: thatServer
-        path: /local/path
-
   # configured wls configuration overrides configmap name
```
Comment (rosemarymarano) [inline model/src/test/resources/oracle/kubernetes/weblogic/domain/v2/domain-sample.yaml]: configmap -> config map

File: docs/domains/Domain.json [hunk 4]
```diff
@@ -232,14 +245,15 @@
           "type": "boolean"
         },
         "webLogicCredentialsSecret": {
+          "description": "The name of a pre-created Kubernetes secret, in the domain\u0027s namepace, that holds the username and password needed to boot WebLogic under the \u0027username\u0027 and \u0027password\u0027 fields.",
```
Comment (rosemarymarano) [inline docs/domains/Domain.json]: needed to boot WebLogic -> needed to boot WebLogic Server

File: docs/domains/index.html [hunk 5]
```diff
@@ -1152,14 +1165,15 @@
           "type": "boolean"
         },
         "webLogicCredentialsSecret": {
+          "description": "The name of a pre-created Kubernetes secret, in the domain\u0027s namepace, that holds the username and password needed to boot WebLogic under the \u0027username\u0027 and \u0027password\u0027 fields.",
```
Comment (rosemarymarano) [inline docs/domains/index.html]: boot WebLogic -> boot WebLogic Server

File: model/src/main/java/oracle/kubernetes/weblogic/domain/v2/BaseConfiguration.java [hunk 6]
```diff
@@ -1,17 +1,18 @@
-// Copyright 2018, Oracle Corporation and/or its affiliates.  All rights reserved.
+// Copyright 2018, 2019, Oracle Corporation and/or its affiliates.  All rights reserved.
 // Licensed under the Universal Permissive License v 1.0 as shown at
 // http://oss.oracle.com/licenses/upl.
 
 package oracle.kubernetes.weblogic.domain.v2;
 
+import com.google.gson.annotations.Expose;
```
Comment (markxnelson) [inline model/src/main/java/oracle/kubernetes/weblogic/domain/v2/BaseConfiguration.java]: is this a new dependency? do we have a PLS record for GSON?

File: model/src/main/java/oracle/kubernetes/weblogic/domain/v2/BaseConfiguration.java [hunk 7]
```diff
@@ -71,14 +64,15 @@ void fillInFrom(BaseConfiguration other) {
     if (other == null) return;
 
     if (serverStartState == null) serverStartState = other.getServerStartState();
-    if (overrideStartPolicyFrom(other)) serverStartPolicy = other.getServerStartPolicy();
+    if (overrideStartPolicyFrom(other)) setServerStartPolicy(other.getServerStartPolicy());
```
Comment (markxnelson) [inline model/src/main/java/oracle/kubernetes/weblogic/domain/v2/BaseConfiguration.java]: you don't have to fix this now, we can do it post-2.0, but we should not have if statements with no braces

File: model/src/main/java/oracle/kubernetes/weblogic/domain/v2/BaseConfiguration.java [hunk 8]
```diff
@@ -111,13 +105,17 @@ void addEnvironmentVariable(String name, String value) {
...
-  public void setServerStartPolicy(String serverStartPolicy) {
-    this.serverStartPolicy = serverStartPolicy;
-  }
+  /**
+   * Tells the operator whether the customer wants the server to be running. For non-clustered
+   * servers - the operator will start it if the policy isn't NEVER. For clustered servers - the
+   * operator will start it if the policy is ALWAYS or the policy is IF_NEEDED and the server needs
+   * to be started to get to the cluster's replica count..
+   *
+   * @since 2.0
```
Comment (markxnelson) [inline model/src/main/java/oracle/kubernetes/weblogic/domain/v2/BaseConfiguration.java]: missing param

File: model/src/main/java/oracle/kubernetes/weblogic/domain/v2/KubernetesResource.java [hunk 9]
```diff
@@ -0,0 +1,88 @@
+// Copyright 2019, Oracle Corporation and/or its affiliates.  All rights reserved.
+// Licensed under the Universal Permissive License v 1.0 as shown at
+// http://oss.oracle.com/licenses/upl.
+
+package oracle.kubernetes.weblogic.domain.v2;
+
+import java.util.*;
```
Comment (markxnelson) [inline model/src/main/java/oracle/kubernetes/weblogic/domain/v2/KubernetesResource.java]: no wildcard imports

File: model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerPod.java [hunk 10]
```diff
@@ -438,19 +355,31 @@ public boolean equals(Object o) {
...
+        .append(Domain.sortOrNull(volumeMounts, VOLUME_MOUNT_COMPARATOR))
         .append(nodeSelector)
         .append(resources)
         .append(podSecurityContext)
         .append(containerSecurityContext)
         .toHashCode();
   }
+
+  private static final Comparator<V1EnvVar> ENV_VAR_COMPARATOR =
+      (a, b) -> {
```
Comment (markxnelson) [inline model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerPod.java]: ooh fancy!

File: model/src/main/java/oracle/kubernetes/weblogic/domain/v2/BaseConfiguration.java (hunk 6 above)
Comment (rjeberhard) [reply model/src/main/java/oracle/kubernetes/weblogic/domain/v2/BaseConfiguration.java]: No, this was in 1.0.  We get this transitively through the kubernetes-client.

File: model/src/main/java/oracle/kubernetes/weblogic/domain/v2/BaseConfiguration.java (hunk 7 above)
Comment (rjeberhard) [reply model/src/main/java/oracle/kubernetes/weblogic/domain/v2/BaseConfiguration.java]: Hmmm...  Good natured disagreement, perhaps.  With the enforced formatting, it's hard to see how braces add to readability when there is just one line.

File: model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerPod.java (hunk 10 above)
Comment (rjeberhard) [reply model/src/main/java/oracle/kubernetes/weblogic/domain/v2/ServerPod.java]: Thanks? :)

**Pull request merged**